### PR TITLE
fix: refactor eventEmailNotifications

### PR DESCRIPTION
**Pull request by Kneesal**
# Description

### Issue

Ability to get the journeyNotification directly from the userTeam/userJourney

[Link to Basecamp Todo](https://3.basecamp.com/3105655/projects)

### Solution

added resolve reference fields to user Journey and user Team resolvers

### CODE REVIEW

File: apps/api-journeys/db/schema.prisma [hunk 1]
```diff
@@ -479,15 +480,16 @@ model JourneysEmailPreference {
...
+  userId                  String
+  journeyId               String
+  journey                 Journey      @relation(fields: [journeyId], references: [id], onDelete: Cascade)
+  userTeamId              String?
+  userJourneyId           String?
+  userTeam                UserTeam?    @relation(fields: [userTeamId], references: [id], onDelete: Cascade)
+  userJourney             UserJourney? @relation(fields: [userJourneyId], references: [id], onDelete: Cascade)
+  visitorInteractionEmail Boolean      @default(false)
 
   @@unique([userId, journeyId])
```
Comment (tataihono) [inline apps/api-journeys/db/schema.prisma:494]:
```suggestion
  @@unique([userId, journeyId])
  @@unique([userId, journeyId, userTeamId])
```

File: apps/api-journeys/db/schema.prisma [hunk 2]
```diff
@@ -479,15 +480,16 @@ model JourneysEmailPreference {
...
-  userTeamId    String?
-  userJourneyId String?
-  userTeam      UserTeam?    @relation(fields: [userTeamId], references: [id], onDelete: Cascade)
-  userJourney   UserJourney? @relation(fields: [userJourneyId], references: [id], onDelete: Cascade)
-  value         Boolean      @default(true)
+model JourneyNotification {
+  id                      String       @id @default(uuid())
+  userId                  String
+  journeyId               String
+  journey                 Journey      @relation(fields: [journeyId], references: [id], onDelete: Cascade)
```
Comment (tataihono) [inline apps/api-journeys/db/schema.prisma:487]: these should all appear at the bottom

File: apps/api-journeys/src/app/modules/journeyNotification/journeyNotification.graphql [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+  id: ID!
+  userId: ID!
+  journeyId: ID!
+  userTeamId: ID
+  userJourneyId: ID
+  visitorInteractionEmail: Boolean!
+}
+
+extend type UserJourney {
+  journeyNotification: [JourneyNotification!]!
```
Comment (tataihono) [inline apps/api-journeys/src/app/modules/journeyNotification/journeyNotification.graphql:11]:
```suggestion
  journeyNotification: JourneyNotification
```

File: apps/api-journeys/src/app/modules/journeyNotification/journeyNotification.graphql [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+}
+
+extend type UserTeam {
+  journeyNotification(journeyId: ID!): [JourneyNotification!]!
```
Comment (tataihono) [inline apps/api-journeys/src/app/modules/journeyNotification/journeyNotification.graphql:15]:
```suggestion
  journeyNotification(journeyId: ID!): JourneyNotification
```

File: apps/api-journeys/src/app/modules/userTeam/userTeam.resolver.ts [hunk 5]
```diff
@@ -128,4 +129,16 @@ export class UserTeamResolver {
   ): Promise<{ __typename: string; id: string }> {
     return { __typename: 'User', id: userTeam.userId }
   }
+
+  @ResolveField('journeyNotification')
+  async journeyNotification(
+    @Parent() userTeam: UserTeam,
+    @Args('journeyId') journeyId: string
+  ): Promise<JourneyNotification[] | null> {
```
Comment (tataihono) [inline apps/api-journeys/src/app/modules/userTeam/userTeam.resolver.ts:137]:
```suggestion
  ): Promise<JourneyNotification | null> {
```

File: apps/api-journeys/src/app/modules/userJourney/userJourney.resolver.ts [hunk 6]
```diff
@@ -257,4 +258,15 @@ export class UserJourneyResolver {
   ): Promise<{ __typename: string; id: string }> {
     return { __typename: 'User', id: userJourney.userId }
   }
+
+  @ResolveField('journeyNotification')
+  async journeyNotification(
+    @Parent() userJourney: UserJourney
+  ): Promise<JourneyNotification[] | null> {
```
Comment (tataihono) [inline apps/api-journeys/src/app/modules/userJourney/userJourney.resolver.ts:265]:
```suggestion
  ): Promise<JourneyNotification | null> {
```

File: apps/api-journeys/src/app/modules/journeyNotification/journeyNotification.resolver.ts [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+  async journeyNotificationsUpdate(
+    @CurrentUserId() userId: string,
+    @Args('input') input: JourneyNotificationUpdateInput
+  ): Promise<JourneyNotification> {
+    const { journeyId } = input
+    return await this.prismaService.journeyNotification.upsert({
+      where: { userId_journeyId: { userId, journeyId } },
+      update: input,
+      create: { userId, ...input }
+    })
```
Comment (tataihono) [inline apps/api-journeys/src/app/modules/journeyNotification/journeyNotification.resolver.ts:35]: you need to find and associate the correct userTeam or userJourney to this journeyNotifcation object.

File: apps/api-journeys/src/app/modules/journeyNotification/journeyNotification.resolver.ts [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+  constructor(private readonly prismaService: PrismaService) {}
+
+  @Query()
+  async journeyNotifications(
+    @Args('journeyId') journeyId: string
+  ): Promise<JourneyNotification[]> {
+    return await this.prismaService.journeyNotification.findMany({
+      where: { journeyId }
+    })
+  }
```
Comment (tataihono) [inline apps/api-journeys/src/app/modules/journeyNotification/journeyNotification.resolver.ts:22]: this bleeds information. the general public could get all the journeyNotifcation data.

**Review by tataihono — CHANGES_REQUESTED** [review]

File: apps/api-journeys/db/schema.prisma [hunk 9]
```diff
@@ -252,14 +252,14 @@ enum UserJourneyRole {
...
-  openedAt                DateTime?
-  EventEmailNotifications EventEmailNotifications[]
+  id                  String                @id @default(uuid())
+  userId              String
+  journeyId           String
+  updatedAt           DateTime              @default(now()) @updatedAt
+  journey             Journey               @relation(fields: [journeyId], references: [id], onDelete: Cascade)
+  role                UserJourneyRole
+  openedAt            DateTime?
+  journeyNotification JourneyNotification[]
```
[the remark this answers is not in the document]
Comment (Kneesal) [reply apps/api-journeys/db/schema.prisma:262]: changed to journeyNotification JourneyNotification?

File: apps/api-journeys/src/app/modules/journeyNotification/journeyNotification.resolver.ts [hunk 10]
```diff
@@ -0,0 +1,45 @@
...
+      ...input,
+      userJourneyId,
+      userTeamId
+    }
+
+    return await this.prismaService.journeyNotification.upsert({
+      where: { userId_journeyId: { userId, journeyId } },
+      update: upsertInput,
+      create: { userId, ...upsertInput }
+    })
```
Comment (tataihono) [inline apps/api-journeys/src/app/modules/journeyNotification/journeyNotification.resolver.ts:43]: this should have a casl guard which ensures that the user has the correct permissions to edit this object.

**Review by tataihono — CHANGES_REQUESTED** [review]

File: apps/api-journeys/src/app/modules/customDomain/customDomain.acl.spec.ts
Comment (tataihono) [inline apps/api-journeys/src/app/modules/customDomain/customDomain.acl.spec.ts:1]: reset

File: apps/api-journeys/src/app/modules/journeyNotification/journeyNotification.resolver.ts [hunk 11]
```diff
@@ -0,0 +1,70 @@
...
+    @CaslAbility() ability: AppAbility,
+    @CurrentUserId() userId: string,
+    @Args('input') input: JourneyNotificationUpdateInput
+  ): Promise<JourneyNotification> {
+    const { journeyId } = input
+    const journey = await this.prismaService.journey.findUnique({
+      where: { id: journeyId },
+      include: { team: { include: { userTeams: true } }, userJourneys: true }
+    })
+    const userJourneyId = journey?.userJourneys?.find(
```
Comment (tataihono) [inline apps/api-journeys/src/app/modules/journeyNotification/journeyNotification.resolver.ts:31]: check user has ability to manage journey.

**Review by tataihono — CHANGES_REQUESTED** [review]

**Review by tataihono — APPROVED** [review]

**Pull request merged**